commerce/domain/inventory/CommerceInventoryServiceSample.cls

```diff
@@ -0,0 +1,81 @@
+public virtual class CommerceInventoryServiceSample extends commerce_inventory.CommerceInventoryService {
+
+    public override commerce_inventory.UpsertReservationResponse upsertReservation(commerce_inventory.UpsertReservationRequest upsertReservationRequest,
+                                                                   commerce_inventory.InventoryReservation currentReservation,
+                                                                   String reservationChangeType) {
+        commerce_inventory.UpsertReservationResponse response = new commerce_inventory.UpsertReservationResponse();
+        response.setSucceed(true);
+        response.setReservationSourceId(upsertReservationRequest.getReservationSourceId());
+        response.setReservationIdentifier(upsertReservationRequest.getReservationIdentifier());
+        List<commerce_inventory.UpsertItemReservationResponse> responseItems = new List<commerce_inventory.UpsertItemReservationResponse>();
+
+        for(commerce_inventory.UpsertItemReservationRequest item : upsertReservationRequest.getItems()) {
+            commerce_inventory.UpsertItemReservationResponse  responseItem = new commerce_inventory.UpsertItemReservationResponse();
+            responseItem.setQuantity(item.getQuantity());
+            responseItem.setReservedAtLocationId(item.getReservedAtLocationId());
+            responseItem.setItemReservationSourceId(item.getItemReservationSourceId());
+            responseItem.setProductId(item.getProductId());
+            responseItems.add(responseItem);
+        }
+
+        response.setItems(responseItems);
+
+        if (upsertReservationRequest.getItems().size() != response.getItems().size()) {
+            throw new NoDataFoundException();
```

Comment:
Please change the error type.  It should be something reflecting a data Conflict.  

Also please assign the boolean expression to a variable to make it super clear what the error condition is: 

eg: 
boolean reservationConflict = upsertReservationRequest.getItems().size() != response.getItems().size(); 
if (reservationConflict) {
 throw new DataConflict("Expected " + x + " got y") 
}
If you need to create a custom execption type: https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_exception_custom.htm